### PR TITLE
[shopsys] moved friendly urls from project-base to the framework

### LINKED ISSUE #3297: migrations Version20240108154625 use attribute redirect_to and redirect_code from project base
Hello,

I would like to report a bug related to the FriendlyUrl entity.

The FriendlyUrl entity has attributes redirectTo and redirectCode in the project base at [FriendlyUrl.php](https://github.com/shopsys/shopsys/blob/v14.0.0/project-base/app/src/Component/Router/FriendlyUrl/FriendlyUrl.php).

However, the migrations in the framework package use these attributes: [Version20240108154625.php](https://github.com/shopsys/shopsys/blob/v14.0.0/packages/framework/src/Migrations/Version20240108154625.php).

This inconsistency causes issues when upgrading to the new version 14.

### PR DESCRIPTION
**Pull request by TomasLudvik**
#### Description, the reason for the PR
Some of the framework functionality was dependent on functionality in the project-base, so this PR moves it to the framework.

#### Fixes issues
closes #3297 

#### Check the appropriate checkboxes below, please

- [x] [BC breaks](https://docs.shopsys.com/en/latest/contributing/backward-compatibility-promise/) 
- [ ] New feature 
- [x] I have read and signed [License Agreement for contributions](https://www.shopsys.com/license-agreement)

----
:globe_with_meridians: Live Preview:
  - https://tl-move-friendly-urls.odin.shopsys.cloud
  - https://cz.tl-move-friendly-urls.odin.shopsys.cloud

### CODE REVIEW

File: packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlData.php [hunk 1]
```diff
@@ -15,4 +15,29 @@ class FriendlyUrlData
...
+
+    /**
+     * @var string|null
+     */
+    public $slug;
+
+    /**
+     * @var int|null
+     */
+    public $entityId;
```
Comment (vitek-rostislav) [inline packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlData.php:22]: could you please remove the original `FriendlyUrlData::$id`? It looks like it is not used anywhere (but the removal should be mentioned in the upgrade notes then anyway)

File: packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlFacade.php [hunk 2]
```diff
@@ -324,4 +330,88 @@ public function findByDomainIdAndSlug(int $domainId, string $slug): ?FriendlyUrl
...
+    {
+        $friendlyUrl = $this->friendlyUrlRepository->findByDomainIdAndSlug($domainId, $slug);
+
+        if ($friendlyUrl === null) {
+            return;
+        }
+
+        $friendlyUrl->setRedirectCode($friendlyUrlData->redirectCode);
+        $friendlyUrl->setRedirectTo($friendlyUrlData->redirectTo);
+        $friendlyUrl->setLastModification(new DateTime('now', new DateTimeZone('UTC')));
```
Comment (vitek-rostislav) [inline packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlFacade.php]: I believe you do not need to set the timezone here, it is UTC by default

File: packages/framework/src/Migrations/Version20240108154624.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Shopsys\FrameworkBundle\Migrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Shopsys\MigrationBundle\Component\Doctrine\Migrations\AbstractMigration;
+
+class Version20240108154624 extends AbstractMigration
```
Comment (vitek-rostislav) [inline packages/framework/src/Migrations/Version20240108154624.php:10]: :ok_hand:

File: packages/framework/src/Model/Blog/Article/BlogArticleDetailFriendlyUrlDataProvider.php [hunk 4]
```diff
@@ -61,18 +60,4 @@ public function getRouteName(): string
     {
         return static::ROUTE_NAME;
     }
-
-    /**
-     * @param int $id
-     * @param string $name
-     * @return \Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlData
-     */
-    protected function createFromIdAndName(int $id, string $name): FriendlyUrlData
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Blog/Article/BlogArticleDetailFriendlyUrlDataProvider.php:70]: this removal should be mentioned in the upgrade notes

File: packages/framework/src/Model/Blog/Category/BlogCategoryDetailFriendlyUrlDataProvider.php [hunk 5]
```diff
@@ -48,26 +47,12 @@ public function getFriendlyUrlData(DomainConfig $domainConfig): array
...
 
         return $friendlyUrlsData;
     }
 
-    /**
-     * @param int $id
-     * @param string $name
-     * @return \Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrlData
-     */
-    public function createFromIdAndName(int $id, string $name): FriendlyUrlData
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Blog/Category/BlogCategoryDetailFriendlyUrlDataProvider.php:62]: this removal should be mentioned in the upgrade notes

**Review by vitek-rostislav — CHANGES_REQUESTED** [review]

**Review by vitek-rostislav — APPROVED** [review]

File: packages/framework/src/Model/Product/ProductFacade.php [hunk 6]
```diff
@@ -368,4 +369,41 @@ public function editProductStockRelation(ProductData $productData, Product $prod
             $productData->productStockData,
         );
     }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Product $product
+     * @param array $originalNames
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Product/ProductFacade.php]:
```suggestion
     * @param string[] $originalNames
```

File: packages/framework/src/Model/Product/ProductFacade.php [hunk 7]
```diff
@@ -368,4 +369,41 @@ public function editProductStockRelation(ProductData $productData, Product $prod
...
+        $this->friendlyUrlFacade->createFriendlyUrls(
+            'front_product_detail',
+            $product->getId(),
+            $changedNames,
+        );
+    }
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Product $product
+     * @param array $originalNames
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Product/ProductFacade.php]:
```suggestion
     * @param string[] $originalNames
```

File: packages/framework/src/Model/Product/ProductFacade.php [hunk 8]
```diff
@@ -368,4 +369,41 @@ public function editProductStockRelation(ProductData $productData, Product $prod
...
+     * @return array
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Product/ProductFacade.php]:
```suggestion
     * @return string[]
```

**Review by vitek-rostislav — APPROVED** [review]

**Comment by vitek-rostislav** [comment]
Hi, check the failing tests, please

**Review by vitek-rostislav — APPROVED** [review]

**Review by pk16011990 — APPROVED** [review]

**Pull request merged**